crates/ruma-events/src/message/content_serde.rs

```diff
@@ -0,0 +1,50 @@
+//! `Serialize` and `Deserialize` implementations for extensible events (MSC1767).
+
+use std::convert::TryFrom;
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use super::{MessageContent, Text};
+
+#[derive(Error, Debug)]
+pub enum MessageContentSerdeError {
+    #[error("missing field `{0}`")]
+    MissingField(String),
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub(crate) struct MessageContentSerDeHelper {
+    /// Plain text short form.
+    #[serde(rename = "org.matrix.msc1767.text", skip_serializing_if = "Option::is_none")]
+    text: Option<String>,
+
+    /// Long form.
+    #[serde(rename = "org.matrix.msc1767.message", default, skip_serializing_if = "Vec::is_empty")]
+    message: Vec<Text>,
+}
+
+impl TryFrom<MessageContentSerDeHelper> for MessageContent {
+    type Error = MessageContentSerdeError;
+
+    fn try_from(helper: MessageContentSerDeHelper) -> Result<Self, Self::Error> {
+        if !helper.message.is_empty() {
+            Ok(Self(helper.message))
+        } else if let Some(text) = helper.text {
+            Ok(Self::plain(text))
+        } else {
+            Err(MessageContentSerdeError::MissingField("m.message".into()))
+        }
+    }
+}
+
+impl From<MessageContent> for MessageContentSerDeHelper {
+    fn from(content: MessageContent) -> Self {
+        let variants = content.variants();
+        if variants.len() == 1 && variants[0].mimetype == "text/plain" {
+            Self { text: Some(variants[0].body.clone()), ..Default::default() }
+        } else {
+            Self { message: variants.to_owned(), ..Default::default() }
+        }
+    }
+}
```

Comment:
Have you tried writing a manual `Serialize` impl for `MessageContent` instead (can be in this same module)? It's often as easy or easier to read, and slightly more efficient.

Reply:
Only `Serialize` or `Deserialize` also?

Reply:
Only `Serialize`, `Deserialize` is rather annoying to write by hand.